### PR TITLE
[core][state] Task backend improve performance

### LINKED ISSUE #32235: [Ray 2.3 Release] `stress_test_many_tasks.json ` `stage_3_time` has 17% regression in 2.3.0
From preliminary performance comparison PR: https://github.com/ray-project/ray/pull/32086/files
* stage_0_time 54% regression
* stage_3_time 17% regression

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
This PR aims to improve performance of the task backend with 3 changes: 
1. Delay conversion of protobuf. We found the protobuf conversion, especially from `TaskSpecification` to `TaskInfoEntry` that's needed for the task metadata has been slow, and was in the critical path of task execution and submission. This PR delays the generation of `rpc::TaskEvnets` before sending in the flush thread. During task execution, it will simply generate a `TaskEvent` entry that's in-memory with a lower overhead. 
2. Fixed the circular buffer that's used as the underlying data structures for the buffered events. This prevents constant resizing when the buffer gets filled up or flushed, which is costly. 
3. Adjust the niceness of the flushing thread, so it has a lower priority than the worker thread. 

## Checks

A summary of release tests run so far, clear signs of improvements on most metrics. While the only one is `stress_test_dead_actors` which has a pretty high variance so not sure how useful it is. 
image

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_event_buffer.cc [hunk 1]
```diff
@@ -53,6 +89,17 @@ Status TaskEventBufferImpl::Start(bool auto_flush) {
     sigaddset(&mask, SIGINT);
     sigaddset(&mask, SIGTERM);
     pthread_sigmask(SIG_BLOCK, &mask, NULL);
+
+    // Decrease the thread priority to allow worker threads to run.
+    int new_nice = std::min(RayConfig::instance().worker_niceness() + 5, 19);
+    new_nice = nice(new_nice);
+    if (new_nice == -1) {
+      RAY_LOG(WARNING) << "Failed to set lower priority for task event buffer io thread: "
```
Comment (cadedaniel) [inline src/ray/core_worker/task_event_buffer.cc:97]: nit: add `strerror`, example: https://github.com/ray-project/ray/pull/30150/files#diff-ee73f8b8b8cf33a817e1d8069d847afba82da19fc7ca011f2b96d2f7a2cac29dR58
nit: include `new_nice` in error message to help to make log more useful

File: src/ray/core_worker/task_event_buffer.h [hunk 2]
```diff
@@ -33,6 +33,45 @@ namespace core {
...
 
+/// A  wrapper class that will be converted to rpc::TaskEvents
+///
+/// This will be created by CoreWorker and stored in TaskEventBuffer, and
+/// when it is being flushed periodically to GCS, it will be converted to
+/// rpc::TaskEvents.
+/// This is an optimization so that converting to protobuf (which is costly)
+/// will not happen in the critical path of task execution/submission.
+class TaskEvent {
+ public:
```
Comment (cadedaniel) [inline src/ray/core_worker/task_event_buffer.h:44]: Can we make these data fields const + populated in a constructor? This is a lot of mutable state..

File: src/ray/core_worker/task_event_buffer.cc [hunk 3]
```diff
@@ -53,6 +89,17 @@ Status TaskEventBufferImpl::Start(bool auto_flush) {
     sigaddset(&mask, SIGINT);
     sigaddset(&mask, SIGTERM);
     pthread_sigmask(SIG_BLOCK, &mask, NULL);
+
+    // Decrease the thread priority to allow worker threads to run.
+    int new_nice = std::min(RayConfig::instance().worker_niceness() + 5, 19);
```
Comment (cadedaniel) [inline src/ray/core_worker/task_event_buffer.cc:94]: nit: move `5` to something like `kEventBufferAdditionalNiceness`?

**Review by cadedaniel — APPROVED** [review]
Nice finds! LGTM, recommend changing the new task event class to be immutable

File: src/ray/core_worker/task_event_buffer.cc [hunk 4]
```diff
@@ -31,8 +68,7 @@ Status TaskEventBufferImpl::Start(bool auto_flush) {
   RAY_CHECK(report_interval_ms > 0)
       << "RAY_task_events_report_interval_ms should be > 0 to use TaskEventBuffer.";
 
-  buffer_.set_capacity(
-      {RayConfig::instance().task_events_max_num_task_events_in_buffer()});
+  buffer_.set_capacity(RayConfig::instance().task_events_max_buffer_size());
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/core_worker/task_event_buffer.cc:71]: 1 M after pointer change right?

File: src/ray/core_worker/task_manager.cc [hunk 5]
```diff
@@ -940,30 +901,17 @@ void TaskManager::RecordTaskStatusEvent(int32_t attempt_number,
...
-    state_updates->set_node_id(node_id->Binary());
-  }
-  if (worker_id.has_value()) {
-    RAY_CHECK(status == rpc::TaskStatus::SUBMITTED_TO_WORKER)
-        << "Worker ID should be included when task status changes to "
-           "SUBMITTED_TO_WORKER.";
-    state_updates->set_worker_id(worker_id->Binary());
-  }
-  gcs::FillTaskStatusUpdateTime(status, absl::GetCurrentTimeNanos(), state_updates);
+  std::unique_ptr<worker::TaskEvent> task_event = std::make_unique<worker::TaskEvent>();
```
Comment (rkooo567) [inline src/ray/core_worker/task_manager.cc:904]: Consider passing data via constructor?

File: src/ray/core_worker/task_event_buffer.cc [hunk 6]
```diff
@@ -14,11 +14,48 @@
...
+
+  // Task info.
+  if (include_task_info) {
+    RAY_CHECK(task_spec != nullptr);
+    gcs::FillTaskInfo(rpc_task_events->mutable_task_info(), *task_spec);
+  }
+
+  if (IsProfileEvent()) {
+    // Profile events.
+    rpc_task_events->mutable_profile_events()->CopyFrom(profile_events.value());
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.cc:39]: is it possible to move or swap this data?

File: src/ray/core_worker/task_event_buffer.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.cc:71]: Now 1M

File: src/ray/core_worker/task_event_buffer.h [hunk 7]
```diff
@@ -33,6 +33,45 @@ namespace core {
...
+  /// Attempt number
+  int32_t attempt_number = -1;
+  /// The task status change if it's a status change event.
+  rpc::TaskStatus task_status = rpc::TaskStatus::NIL;
+  /// The time when the task status change happens.
+  int64_t timestamp = -1;
+  /// Pointer to the task spec.
+  std::shared_ptr<const TaskSpecification> task_spec = nullptr;
+  /// If task info needs to be included in rpc::TaskEvents
+  bool include_task_info = false;
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:66]: maybe we don't need this? (just check if task_spec is null?)

File: src/ray/core_worker/task_event_buffer.h [hunk 8]
```diff
@@ -33,6 +33,45 @@ namespace core {
...
+/// will not happen in the critical path of task execution/submission.
+class TaskEvent {
+ public:
+  /// Convert itself a rpc::TaskEvents
+  ///
+  /// \param[out] rpc_task_events The rpc task event to be filled.
+  void ToRpcTaskEvents(rpc::TaskEvents *rpc_task_events) const;
+
+  /// If it is a profile event.
+  bool IsProfileEvent() const { return profile_events.has_value(); }
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:51]: Consider breaking the class into ProfileEvent & StatusUpdateEvent?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
LGTM

**Comment by rkooo567** [comment]
Does it include vector change now?

**Comment by cadedaniel** [comment]
Can we merge this ASAP? This and one other PR are on the critical path for running release tests before handing off for product testing.

**Pull request merged**